src/defrag.c

```diff
@@ -840,299 +914,450 @@ int defragLaterItem(dictEntry *de, unsigned long *cursor, long long endtime, int
-                start_stat = server.stat_active_defrag_hits;
-            }
+    server.stat_total_active_defrag_time += elapsedUs(server.stat_last_active_defrag_time);
+    server.stat_last_active_defrag_time = 0;
+    server.active_defrag_cpu_percent = 0;
+}
+
+
+/* Must be called at the start of the timeProc as it measures the delay from the end of the previous
+ * timeProc invocation when performing the computation. */
+static int computeDefragCycleUs(void) {
+    long dutyCycleUs;
+
+    int targetCpuPercent = server.active_defrag_cpu_percent;
+    serverAssert(targetCpuPercent > 0 && targetCpuPercent < 100);
 
-            db = &server.db[current_db];
-            kvstoreDictLUTDefrag(db->keys, dictDefragTables);
-            kvstoreDictLUTDefrag(db->expires, dictDefragTables);
-            defrag_stage = 0;
-            defrag_cursor = 0;
-            slot = -1;
-            defrag_later_item_in_progress = 0;
+    static int prevCpuPercent = 0; // STATIC - this persists
+    if (targetCpuPercent != prevCpuPercent) {
+        /* If the targetCpuPercent changes, the value might be different from when the last wait
+         *  time was computed.  In this case, don't consider wait time.  (This is really only an
+         *  issue in crazy tests that dramatically increase CPU while defrag is running.) */
+        defrag.timeproc_end_time = 0;
+        prevCpuPercent = targetCpuPercent;
+    }
+
+    // Given when the last duty cycle ended, compute time needed to achieve the desired percentage.
+    if (defrag.timeproc_end_time == 0) {
+        // Either the first call to the timeProc, or we were paused for some reason.
+        defrag.timeproc_overage_us = 0;
+        dutyCycleUs = server.active_defrag_cycle_us;
+    } else {
+        long waitedUs = getMonotonicUs() - defrag.timeproc_end_time;
+        /* Given the elapsed wait time between calls, compute the necessary duty time needed to
+         *  achieve the desired CPU percentage.
+         *  With:  D = duty time, W = wait time, P = percent
+         *  Solve:    D          P
+         *          -----   =  -----
+         *          D + W       100
+         *  Solving for D:
+         *     D = P * W / (100 - P)
+         *
+         * Note that dutyCycleUs addresses starvation.  If the wait time was long, we will compensate
+         *  with a proportionately long duty-cycle.  This won't significantly affect perceived
+         *  latency, because clients are already being impacted by the long cycle time which caused
+         *  the starvation of the timer. */
+        dutyCycleUs = targetCpuPercent * waitedUs / (100 - targetCpuPercent);
+
+        // Also adjust for any accumulated overage(underage).
+        dutyCycleUs -= defrag.timeproc_overage_us;
+        defrag.timeproc_overage_us = 0;
+
+        if (dutyCycleUs < server.active_defrag_cycle_us) {
+            /* We never reduce our cycle time, that would increase overhead.  Instead, we track this
```

Comment:
The inter-cycle delay is computed in milliseconds, and is not accurate as it may be delayed.  At this point in the code, we are computing the number of micro-seconds needed to achieve the desired CPU percentage since the last cycle.  The actual computed number may be a little above/below the configured duty-cycle-us.  If we require less CPU than the duty-cycle, we don't reduce the duty-cycle.  Instead, we add to the "overage" so that we can later compensate with an extra millisecond delay when we get enough overage.

On the flip size, if the computed dutyCycleUs is greater than the configured/intended `server.active_defrag_cycle_us`, that means that we are being starved, and that the timer didn't fire when it was supposed to.  In this case, we allow the duty-cycle to run longer, to compensate for the starvation.